econml/sklearn_extensions/linear_model.py

```diff
@@ -1905,3 +1905,132 @@ def model_gen(y):
             self._param_var = np.array([mdl.cov_params() for mdl in self.models])
 
         return self
+
+
+class StatsModels2SLS(_StatsModelsWrapper):
+    """
+    Class that solves the moment equation E[(y-theta*T)*Z]=0
+
+    Parameters
+    ----------
+    cov_type : one of {'HC0', 'HC1', 'nonrobust' or None} (optional, default='HC0')
+        Indicates how the covariance matrix is estimated.
+    """
+
+    def __init__(self, cov_type="HC0"):
+        self.fit_intercept = False
+        self.cov_type = cov_type
+        return
+
+    def _check_input(self, Z, T, y, sample_weight):
+        """Check dimensions and other assertions."""
+
+        # set default values for None
+        if sample_weight is None:
+            sample_weight = np.ones(y.shape[0])
+
+        # check array shape
+        assert (T.shape[0] == Z.shape[0] == y.shape[0] == sample_weight.shape[0]), "Input lengths not compatible!"
+
+        # weight X and y
+        weighted_Z = Z * np.sqrt(sample_weight).reshape(-1, 1)
+        weighted_T = T * np.sqrt(sample_weight).reshape(-1, 1)
+        if y.ndim < 2:
+            weighted_y = y * np.sqrt(sample_weight)
+        else:
+            weighted_y = y * np.sqrt(sample_weight).reshape(-1, 1)
+        return weighted_Z, weighted_T, weighted_y
+
+    def fit(self, y, T, Z, sample_weight=None, freq_weight=None, sample_var=None):
+        """
+        Fits the model.
+
+        Parameters
+        ----------
+        y :  {(N,), (N, p)} nd array like
+            output variables
+        T :  {(N, p)} nd array like
+            treatment variables
+        Z :  {(N, p)} nd array like
+            instrumental variables
+        sample_weight : (N,) array like or None
+            Individual weights for each sample. If None, it assumes equal weight.
+        freq_weight: (N, ) array like of integers or None
+            Weight for the observation. Observation i is treated as the mean
+            outcome of freq_weight[i] independent observations.
+            When ``sample_var`` is not None, this should be provided.
+        sample_var : {(N,), (N, p)} nd array like or None
+            Variance of the outcome(s) of the original freq_weight[i] observations that were used to
+            compute the mean outcome represented by observation i.
+
+
+        Returns
+        -------
+        self : StatsModels2SLS
+        """
+        assert freq_weight is None, "freq_weight is not supported yet for this class!"
+        assert sample_var is None, "sample_var is not supported yet for this class!"
+
+        Z, T, y = self._check_input(Z, T, y, sample_weight)
+
+        self._n_out = 0 if y.ndim < 2 else y.shape[1]
+
+        # solve moment equation E[(y-theta*T)*Z]=0
+        # theta = (Z.T*T)^(-1)*Z.T*y
+        zT_t = np.dot(Z.T, T)
+        zT_y = np.dot(Z.T, y)
+        param = np.linalg.solve(zT_t, zT_y)
```

Comment:
I think it only works with `d_T==d_Z`? we need to compute the inverse of (Z.T*T), it needs to be a square matrix? I will check they have same dimension.

Reply:
2SLS can work even beyond since you first do the projection of Z on T, which makes the projected Z have the same dimesnion as T. 

In your 2SLS you find the point estimate by direct inversion of the matrix, but then you find the covariance using the two stage projection.

If you also use the two stage projection for the point estimate it would also work for Z larger than T.

Reply:
See for isntance how statsmodels 2SLS fit does it:
https://www.statsmodels.org/stable/_modules/statsmodels/sandbox/regression/gmm.html#IV2SLS.fit

Reply:
Done!